File: WoodPeckerDemo/WoodPeckerDemo/Modules/Notification/NotificationViewController.m
//
//  NotificationViewController.m
//  WoodPeckerDemo
//
//  Created by xiaogang zhang on 2018/6/4.
//  Copyright © 2018年 lifebetter. All rights reserved.
//

#import "NotificationViewController.h"
@import UserNotifications;

@interface NotificationViewController ()

@end

@implementation NotificationViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (IBAction)fireButtonPressed:(id)sender {
    if (@available(iOS 10.0, *)) {
        UNUserNotificationCenter * notiCenter = [UNUserNotificationCenter currentNotificationCenter];
        [notiCenter requestAuthorizationWithOptions:(UNAuthorizationOptionAlert|UNAuthorizationOptionBadge|UNAuthorizationOptionSound) completionHandler:^(BOOL granted, NSError * _Nullable error) {
            if(granted){
                UNUserNotificationCenter * notiCenter = [UNUserNotificationCenter currentNotificationCenter];
                NSString * identifier = [NSString stringWithFormat:@"%d",arc4random()];
                NSString * requestIdentifier = [NSString stringWithFormat:@"request identifier%@",identifier];
                UNMutableNotificationContent * content = [[UNMutableNotificationContent alloc] init];
                content.badge = [NSNumber numberWithInteger:arc4random()%10];
                content.body = [NSString stringWithFormat:@"Thanks for using woodpecker %@",identifier];
                content.title = [NSString stringWithFormat:@"This is title %@",identifier];
                content.subtitle = [NSString stringWithFormat:@"This is subtitle %@",identifier];
                content.userInfo = @{
                                     @"identifier" : identifier,
                                     };
                content.categoryIdentifier = @"actionCategory";
                UNTimeIntervalNotificationTrigger * trigger = [UNTimeIntervalNotificationTrigger triggerWithTimeInterval:3 repeats:NO];
                UNNotificationRequest * request = [UNNotificationRequest requestWithIdentifier:requestIdentifier content:content trigger:trigger];
                [notiCenter addNotificationRequest:request withCompletionHandler:^(NSError * _Nullable error) {
                    if(error){
                        NSLog(@"add local notification error: %@",error);
                    }else{
                        NSLog(@"add local notification success");
                    }
                }];
            }else{
                NSLog(@"not granted. error: %@",error);
            }
        }];
    } else {
        NSString *message = @"Sorry, this tool only support iOS 10+";
        UIAlertController * alert = [UIAlertController alertControllerWithTitle:@"Alert" message:message preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction * confirm = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:nil];
        [alert addAction:confirm];
        [self presentViewController:alert animated:YES completion:nil];
    }
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

@end

File: WoodpeckerMacDemo/WoodpeckerMacDemo/Pages/Notification/NotificationViewController.swift
//
//  NotificationViewController.swift
//  WoodpeckerMacDemo
//
//  Created by xiaogang zhang on 2019/6/1.
//  Copyright © 2019 LIFE BETTER. All rights reserved.
//

import Cocoa
import UserNotifications

class NotificationViewController: NSViewController, UNUserNotificationCenterDelegate {

    override func viewDidLoad() {
        super.viewDidLoad()
        setup()
    }
    
    func setup() {
        let notiCenter = UNUserNotificationCenter.current()
        notiCenter.delegate = self
        let openAction = UNNotificationAction.init(identifier: UNNotificationDefaultActionIdentifier, title: "Okay", options: UNNotificationActionOptions.init(rawValue: 0))
        let category = UNNotificationCategory.init(identifier: "category_id", actions: [openAction], intentIdentifiers: [], options: UNNotificationCategoryOptions.init(rawValue: 0))
        let set = Set.init([category])
        notiCenter.setNotificationCategories(set)
    }
    
    @IBAction func scheduleButtonClicked(_ sender: Any) {
        let notiCenter = UNUserNotificationCenter.current()
        notiCenter.requestAuthorization(options: [UNAuthorizationOptions.alert,UNAuthorizationOptions.badge,UNAuthorizationOptions.sound]) { (okay, error) in
            if okay {
                self.schedule()
            }
            if let err = error {
                print(err)
                let alert = NSAlert.init()
                alert.messageText = "request authorization failed, you can try in your own project later"
                alert.addButton(withTitle: "Okay")
                alert.runModal();
            }
        }
    }
    
    func schedule () {
        let notiCenter = UNUserNotificationCenter.current()
        let content = UNMutableNotificationContent.init()
        content.badge = 3
        content.body = "I'm body"
        content.title = "I'm title"
        content.subtitle = "I'm subtitle"
        content.userInfo = [
            "key" : "value"
        ]
        content.categoryIdentifier = "category_id";
        //60s
        let trigger = UNTimeIntervalNotificationTrigger.init(timeInterval: 60, repeats: true)
        let identifier = String(arc4random())
        let request = UNNotificationRequest.init(identifier: identifier, content: content, trigger: trigger)
        notiCenter.add(request) { (error) in
            
        }
    }
    
    public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        
    }
    
    @IBAction func removeAllButtonClicked(_ sender: Any) {
        let notiCenter = UNUserNotificationCenter.current()
        notiCenter.removeAllPendingNotificationRequests();
    }
}
